src/level_set/fortran/levelsetops2d.f.m4

```diff
@@ -934,7 +1071,302 @@ c     Compute first order derivatives
+     &            (V(i0,i1))*h2**2)/(h1*h2*(h1 + h2))
+
+            h1 = hy - hym; h2 = hym
+            Dyb = (V(i0,i1)*h1**2 + two*V(i0,i1)*h1*h2 +
+     &            V(i0,i1-1)*h2**2)/(h1*h2*(h1 + h2))
+          endif
+
+c         Compute first order derivatives
+
+          Dxm = (one - wxm)*Dxc + wxm*Dxl
+          Dxp = (one - wxp)*Dxc + wxp*Dxr
+
+          Dym = (one - wym)*Dyc + wym*Dyb
+          Dyp = (one - wyp)*Dyc + wyp*Dyt
+
+          H(i0,i1) = HG(Dxp,Dxm,Dyp,Dym,sgn)
+        enddo
+      enddo
+
+      return
+      end
+c
+cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+c     Mass constraint on level set to ensure that it does not lose volume
+c
+ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
+c
+      subroutine projectlsmassconstraint2d(
+     &     U,U_gcw,
+     &     C,C_gcw,
+     &     V,V_gcw,
+     &     H,H_gcw,
+     &     ilower0,iupper0,
+     &     ilower1,iupper1,
+     &     dx)
+c
+      implicit none
+include(TOP_SRCDIR/src/fortran/const.i)dnl
 
+c
+c     Input.
+c
+      INTEGER ilower0,iupper0
+      INTEGER ilower1,iupper1
+      INTEGER U_gcw,C_gcw,V_gcw,H_gcw
+
+c
+c     Input/Output.
+c
+      REAL U(CELL2d(ilower,iupper,U_gcw))
+      REAL C(CELL2d(ilower,iupper,C_gcw))
+      REAL V(CELL2d(ilower,iupper,V_gcw))
+      REAL H(CELL2d(ilower,iupper,H_gcw))
+      REAL dx(0:NDIM-1)
+c
+c     Local variables.
+c
+      INTEGER i0,i1
+      REAL    hx,hy,hmin
+      REAL    lambda
+      REAL    hij,hmn,phi0
+      INTEGER m,n
+      REAL    nmr,dnr
+      REAL    w(-1:1,-1:1)
+
+      hx = dx(0)
+      hy = dx(1)
+      hmin = dmin1(hx,hy)
+
+c     Compute integration weights based on Simpson's rule
+      do m = -1,1
+        do n = -1,1
+          if(m * n .eq. 0) then
+            w(m,n) = 10.d0
+          else
+            w(m,n) = 1.d0
+          endif
+         enddo
+       enddo
+       w(0,0) = 0.d0
+
+      do i1 = ilower1,iupper1
+        do i0 = ilower0,iupper0
+            phi0 = V(i0,i1)
+            if (abs(phi0) .gt. hmin) then
+              hij = zero
+            else
+              hij = one/(two*hmin) * (one+cos(pi*phi0/hmin))
+            endif
+            nmr = 100.d0*hij*(C(i0,i1) - phi0)
+            dnr = 100.d0*(hij**2)*H(i0,i1)
+            do m = -1,1
+              do n = -1,1
+                phi0 = V(i0+m,i1+n)
+                if (abs(phi0) .gt. hmin) then
+                  hmn = zero
+                else
+                  hmn = one/(two*hmin) * (one+cos(pi*phi0/hmin))
+                endif
+                nmr = nmr + w(m,n)*hmn*(C(i0+m,i1+n) - phi0)
+                dnr = dnr + w(m,n)*hmn**2*H(i0+m,i1+n)
+              enddo
+            enddo
+            nmr = hx*hy/144.d0*nmr
+            dnr = hx*hy/144.d0*dnr
+            lambda = -nmr/dnr
+
+            if (dnr .gt. zero) then
+              U(i0,i1) = C(i0,i1) + lambda*hij*H(i0,i1)
+            endif
+
```

Comment:
Is there a way to store lambda? If lambda can be back-propagated to RelaxationLSMethod, one can basically print the norm of lambda. It should go down to zero if the mass constraint is proper and the solution has converged. One can plog this information.